Libs/Mesh/Mesh.cpp

```diff
@@ -967,6 +988,600 @@ std::ostream& operator<<(std::ostream &os, const Mesh& mesh)
+
+    // Setting scalar field for value
+    this->setField("value", values);
+
+    return absvalues;
+}
+
+std::vector<Eigen::Matrix3d> Mesh::setGradientFieldForFFCs(vtkSmartPointer<vtkDoubleArray> absvalues, Eigen::MatrixXd V, Eigen::MatrixXi F){
+    // Definition of gradient field
+    vtkSmartPointer<vtkDoubleArray> vf = vtkSmartPointer<vtkDoubleArray>::New();
+    vf->SetNumberOfComponents(3);
+    vf->SetNumberOfTuples(this->mesh->GetNumberOfPoints());
+    vf->SetName("vf");
+
+    // *Computing gradients for FFCs for each face
+    // Compute gradient operator
+    //std::cout << "Gradient preprocessing like in Karthik's code" << std::endl;
+    Eigen::SparseMatrix<double> G;
+    igl::grad(V, F, G);
+    // Flattened version of libigl's gradient operator
+    std::vector<Eigen::Matrix3d> face_grad_;
+
+    Eigen::MatrixXd grads(this->mesh->GetNumberOfPoints(), 3);
+    grads.fill(0.);
+
+    // Flatten the gradient operator so we can quickly compute the gradient at a given point
+    face_grad_.resize(F.rows());
+    size_t n_insertions = 0;
+    for(int k=0; k<G.outerSize(); k++) {
+      for(Eigen::SparseMatrix<double>::InnerIterator it(G, k); it; ++it) {
+        const double val = it.value();
+        const auto r = it.row();
+        const auto c = it.col();
+
+        const auto f = r % F.rows();
+        const auto axis = r / F.rows();
+        for(int i=0; i<3; i++) {
+          if(F(f, i) == c) {
+            face_grad_[f](axis, i) = val;
+            n_insertions++;
+            break;
+          }
+        }
+      }
+    }
+
+    //std::cout << "Computing grads" << std::endl;
+
+    // Definition of gradient field for faces
+    vtkSmartPointer<vtkDoubleArray> vff = vtkSmartPointer<vtkDoubleArray>::New();
+    vff->SetNumberOfComponents(3);
+    vff->SetNumberOfTuples(this->mesh->GetNumberOfCells());
+    vff->SetName("vff");
+
+    // Computes grad vec for each face
+    for(size_t i = 0; i < F.rows(); i++){
+        const Eigen::Vector3d vert_dists(absvalues->GetValue(F(i,0)), absvalues->GetValue(F(i,1)), absvalues->GetValue(F(i,2)));
+
+        // Compute gradient of geodesics
+        const auto& G = face_grad_[i];
+        Eigen::Vector3d out_grad_eigen = (G*vert_dists).rowwise().sum();
+        grads.row(F(i,0)) += out_grad_eigen;
+        grads.row(F(i,1)) += out_grad_eigen;
+        grads.row(F(i,2)) += out_grad_eigen;
+        vff->SetTuple3(i, out_grad_eigen(0), out_grad_eigen(1), out_grad_eigen(2));
+        //out_grad_eigen *= geo_dist / out_grad_eigen.norm();
+    }
+
+    // Setting gradient field
+    for(size_t i = 0; i < this->mesh->GetNumberOfPoints(); i++){
+        vf->SetTuple3(i, grads(i,0), grads(i,1), grads(i,2));
+    }
+
+    this->setField("Gradient", vf);
+
+    this->mesh->GetCellData()->AddArray(vff);
+
+    return face_grad_;
+}
+
+void Mesh::visualizeVectorFieldForFFCs(vtkSmartPointer<vtkDoubleArray> values, std::vector<Eigen::Matrix3d> face_grad_, Eigen::MatrixXd V, Eigen::MatrixXi F){
```

Comment:
I think this belongs elsewhere and not in the Mesh class.  I don't think the Mesh class needs the ability to visualize itself.  This could go in MeshUtils or a new MeshVizUtils or something.